pipelines/get_wafer_offset.py

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python
+
+# Copyright (c) 2020 Simons Observatory.
+# Full license can be found in the top level "LICENSE" file.
+
+import argparse
+import os
+import sys
+from collections import OrderedDict
+
+import healpy as hp
+import numpy as np
+import toast.qarray as qa
+
+import sotodlib.sim_hardware as hardware
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="This program measures the median offset of subset of "
+        "detectors from boresight.",
+        usage="get_wafer_offset [options] (use --help for details)")
+
+    group = parser.add_mutually_exclusive_group(required=True)
+    group.add_argument(
+        "--tube_slots",
+        help="Comma-separated list of optics tube slots: c1 (UHF), i5 (UHF), "
+        " i6 (MF), i1 (MF), i3 (MF), i4 (MF), o6 (LF).  "
+    )
+    group.add_argument(
+        "--wafer_slots",
+        help="Comma-separated list of optics tube slots. "
+    )
+    parser.add_argument(
+        "--reverse", action="store_true", help="Reverse offsets")
+
+    args = parser.parse_args()
+
+    hw = hardware.get_example()
+
+    # Which telescope?
+
+    if args.wafer_slots is not None:
+        wafer_slots = args.wafer_slots.split(",")
+        wafer_map = hw.wafer_map()
+        tube_slots = [wafer_map["tube_slots"][ws] for ws in wafer_slots]
+    else:
+        tube_slots = args.tube_slots.split(",")
+
+    telescope = None
+    for tube_slot in tube_slots:
+        for telescope_name, telescope_data in hw.data["telescopes"].items():
+            if tube_slot in telescope_data["tube_slots"]:
+                if telescope is None:
+                    telescope = telescope_name
+                elif telescope != telescope.name:
+                    raise RuntimeError(
+                        f"Tubes '{tube_slots}' span more than one telescope"
+                    )
+        if telescope is None:
+            raise RuntimeError(
+                f"Failed to match tube_slot = '{tube_slot}' with a telescope"
+            )
+
+    # Which detectors?
+
+    hw.data["detectors"] = hardware.sim_telescope_detectors(hw, telescope)
+
+    match = {}
+    tube_slots = None
+    if args.wafer_slots is not None:
+        match["wafer_slot"]  = args.wafer_slots.split(",")
+    elif args.tube_slots is not None:
+        tube_slots = args.tube_slots.split(",")
+
+    hw = hw.select(tube_slots=tube_slots, match=match)
+    ndet = len(hw.data["detectors"])
+
+    # print(f"tube_slots = {tube_slots}, match = {match} leaves {ndet} detectors")
+
+    # Average detector offset
```

Comment:
Note that I do a duplicate calculation in the visualization tools (https://github.com/simonsobs/sotodlib/blob/f560e3196c36ebdd890516e58bac94da24f7708d/sotodlib/vis_hardware.py#L108).  We should make a function that does this and can be used in both places.  That can wait until after this is merged, but let's not forget about it.

Reply:
Agreed. We don't want duplicated code. The context for this pipeline is that we needed to automate the offset estimation for shell scripting.